### PR TITLE
migrate to chart.js 3

### LINKED ISSUE #180: Compatibility with Chart js 3.0
Do you have any plans to add next version compatibile with Chart.Js@next(3.0)?

### PR DESCRIPTION
**Pull request by sgratzl**
all examples seem to work and look the same with  3.0.0-alpha

closes #180

### CODE REVIEW

**Comment by simonbrunel** [comment]
Thanks a lot @sgratzl for tackling this migration. Not sure what's going on with Travis, something wrong with tooltips. I will try to review it asap and checkout your branch to make sure that everything works as expected.

@kurkle could you please have a look to that PR since I'm not familiar with changes in v3.

File: docs/guide/options.md [hunk 1]
```diff
@@ -12,12 +12,12 @@ The following table lists all available options:
...
 | [`clamp`](positioning.md#clamping) | `boolean` | Yes | Yes | `false`
 | [`clip`](positioning.md#clipping) | `boolean` | Yes | Yes | `false`
-| `color` | [`Style`](#style-options) | Yes | Yes | [`defaultFontColor`](http://www.chartjs.org/docs/latest/general/fonts.html)
+| `color` | [`Style`](#style-options) | Yes | Yes | [`fontColor`](http://www.chartjs.org/docs/latest/general/fonts.html)
 | [`display`](positioning.md#visibility) | `boolean` \| `string` | Yes | Yes | `true`
 | `font` | `object` | Yes | Yes | -
-| `font.family` | `string` | - | - | [`defaultFontFamily`](http://www.chartjs.org/docs/latest/general/fonts.html)
-| `font.size` | `string` | - | - | [`defaultFontSize`](http://www.chartjs.org/docs/latest/general/fonts.html)
-| `font.style` | `string` | - | - | [`defaultFontStyle`](http://www.chartjs.org/docs/latest/general/fonts.html)
+| `font.family` | `string` | - | - | [`fontFamily`](http://www.chartjs.org/docs/latest/general/fonts.html)
```
Comment (kurkle) [inline docs/guide/options.md:18]: These have changed further in chart.js master (to font.family etc)
But this is correct for the first aplha

File: samples/utils.js [hunk 2]
```diff
@@ -2,7 +2,7 @@
 
 (function(global) {
 	var Samples = global.Samples || (global.Samples = {});
-	var Color = global.Color;
+	var Color = global.Chart ? global.Chart.helpers.color : function() {};
```
Comment (kurkle) [inline samples/utils.js:5]: @simonbrunel does this one need to work when there is no `global.Chart`?

File: test/specs/events.spec.js [hunk 3]
```diff
@@ -1,6 +1,13 @@
 import Chart from 'chart.js';
 
 describe('events', function() {
+	beforeEach(() => {
+		Chart.helpers.requestAnimFrameBak = Chart.helpers.requestAnimFrame;
+		Chart.helpers.requestAnimFrame = (cb) => cb();
```
Comment (kurkle) [inline test/specs/events.spec.js:6]: Not quite sure why this is needed?

Comment (sgratzl) [reply test/specs/events.spec.js:6]: as far as I understood, the chart.js animations use `requestAnimFrame` to render even the first frame. Thus, it renders asynchronously. However, the tests are synchronous, thus the patch to fix it.

File: samples/utils.js (hunk 2 above)
Comment (sgratzl) [reply samples/utils.js:5]: the `utils.js` is also imported in the generic `index.html` page which hasn't chart.js included. Thus the check

File: test/specs/events.spec.js (hunk 3 above)
Comment (kurkle) [reply test/specs/events.spec.js:6]: Ok, that could be considered a bug in chart.js
Another option could be to disable animations.

**Comment by stockiNail** [comment]
I had a look to the PR and it doesn't look like working with CHART.js 3-beta.
FYI I have forked and changed to use the master of CHART.JS and it sounds working (I didn't change the tests and docs).

**Comment by simonbrunel** [comment]
@stockiNail @sgratzl I realize that I have pending review comments (since a few months ago) that I never submitted because the CI is failing. Can one of you update this PR against the latest Chart.js version to get the CI passing so I can finish to review it?

**Comment by stockiNail** [comment]
@simonbrunel I'm quite busy at the moment on other stuff. I'm gonna try to take the the @sgratzl branch and update it as soon as I can.

**Comment by stockiNail** [comment]
> @stockiNail @sgratzl I realize that I have pending review comments (since a few months ago) that I never submitted because the CI is failing. Can one of you update this PR against the latest Chart.js version to get the CI passing so I can finish to review it?

@sgratzl @simonbrunel as promised, I have submitted a PR (**[https://github.com/sgratzl/chartjs-plugin-datalabels/pull/1](https://github.com/sgratzl/chartjs-plugin-datalabels/pull/1)**) into @sgratzl fork. 

Please have a look, also to the description on PR because I'm using **[dist/master](https://www.chartjs.org/dist/master/chart.js)** of CHART.JS (different from 3.0.0-beta.4, current last pre-release).

**Comment by santam85** [comment]
Opened https://github.com/sgratzl/chartjs-plugin-datalabels/pull/4 for a few fixes to typings and unit-tests. Is there a procedure to see diffs with the comparison files, and/or regenerate them?

**Comment by santam85** [comment]
Created another: https://github.com/sgratzl/chartjs-plugin-datalabels/pull/5, with more unit-test and linting fixes.

**Comment by santam85** [comment]
@simonbrunel need some guidance to fix the remaining unit tests. It looks like it's going to be near impossible to provide a fix due to the code at https://github.com/chartjs/Chart.js/blob/afbe03cce4df724fab04eafda0a7933f42a186a0/src/controllers/controller.bar.js#L436

The bar values get limited in the controller to some arbitrary values just outside of the chart area, making it impossible to position the labels accurately with the data available on the element scope.

Shall I create a PR to the Bar controller to expose that info on the elements, or do you know of an alternative way?

**Comment by simonbrunel** [comment]
@santam85 I'm not sure to fully get what is broken (and when), can you post some use cases & screenshots that show the difference between before and now? Also the failing tests output images & diff would probably help to understand.

> Shall I create a PR to the Bar controller to expose that info on the elements, or do you know of an alternative way?

I think knowing the exact element (pixel) geometry is quite vital to write such plugin (the internal clipping made by the bar controller should however stay private since it's implementation details). So I would probably suggest a better API in the core lib. Unfortunately, I don't think of an alternative solution (@kurkle?)

**Comment by santam85** [comment]
It's breaking when positioning labels, for scales that have min/max values excluding significant portions of the bars:
Screenshot 2020-11-06 at 11 52 33
Before there was a 1-1 relationship between values and bar sizes, so labels would be positioned to the effective pixel position if using start/end/center, would just be clipped by the chart area if outside the visible range. Now start/end are effectively clamped by the bar controller to +-10px above/belo the chart area, skewing the centerpoint calculation too.
Fixing it without changing Chart.js would mean replicating a lot of the logic present in the bar controller, which I believe it's not something we should do.

**Comment by kurkle** [comment]
The +/-10px workaround might be already obsolete, because the pixel values returned by scales are today sanitized to values that do not break canvas rendering.

**Comment by sgratzl** [comment]
closing in favor of #202

**Pull request closed without merging**